internal/impl/sftp/input.go

```diff
@@ -173,177 +176,243 @@ func newSFTPReaderFromParsed(conf *service.ParsedConfig, mgr *service.Resources)
+			s.stateLock.Lock()
+			s.closeScanner(ctx)
+			s.closeClient(false)
+			s.stateLock.Unlock()
+			err = service.ErrNotConnected
+		}
+		return nil, nil, err
+	}
+	return parts, codecAckFn, nil
+}
 
-	if scanner == nil || client == nil {
-		return nil, nil, service.ErrNotConnected
+func (s *sftpReader) tryReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	if err := s.initScanner(ctx); err != nil {
+		return nil, nil, err
 	}
 
-	parts, codecAckFn, err := scanner.NextBatch(ctx)
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	parts, codecAckFn, err := s.scanner.NextBatch(ctx)
 	if err != nil {
 		if ctx.Err() != nil {
 			return nil, nil, ctx.Err()
 		}
-		_ = scanner.Close(ctx)
-		s.scannerMut.Lock()
-		if s.currentPath == currentPath {
-			s.scanner = nil
-			s.currentPath = ""
-		}
-		s.scannerMut.Unlock()
+		_ = s.scanner.Close(ctx)
+		s.scanner = nil
+		s.currentPath = ""
 		if errors.Is(err, io.EOF) {
 			err = service.ErrNotConnected
 		}
 		return nil, nil, err
 	}
 
 	for _, part := range parts {
-		part.MetaSetMut("sftp_path", currentPath)
+		part.MetaSetMut("sftp_path", s.currentPath)
 	}
 
-	return parts, func(ctx context.Context, res error) error {
-		return codecAckFn(ctx, res)
-	}, nil
+	return parts, codecAckFn, nil
 }
 
-func (s *sftpReader) Close(ctx context.Context) error {
-	s.scannerMut.Lock()
+func (s *sftpReader) initScanner(ctx context.Context) error {
+	s.stateLock.Lock()
 	scanner := s.scanner
-	s.scanner = nil
-	client := s.client
-	s.client = nil
-	s.paths = nil
-	s.scannerMut.Unlock()
-
+	s.stateLock.Unlock()
 	if scanner != nil {
-		if err := scanner.Close(ctx); err != nil {
-			s.log.With("error", err).Warn("Failed to close consumed file")
+		return nil
+	}
+
+	var file *sftp.File
+	var path string
+	for {
+		var ok bool
+		var err error
+		path, ok, err = s.pathProvider.Next(ctx, s.borrowClient)
+		if err != nil {
+			return fmt.Errorf("finding next file path: %w", err)
+		}
+		if !ok {
+			return service.ErrEndOfInput
+		}
+
+		client, releaseClient, err := s.borrowClient()
+		if err != nil {
+			return err
+		}
+		file, err = client.Open(path)
+		releaseClient()
+		if err != nil {
+			s.log.With("path", path, "err", err.Error()).Warn("Unable to open previously identified file")
+			if os.IsNotExist(err) {
+				// If we failed to open the file because it no longer exists then we
+				// can "ack" the path as we're done with it. Otherwise we "nack" it
+				// with the error as we'll want to reprocess it again later.
+				err = nil
+			}
+			if ackErr := s.pathProvider.Ack(ctx, path, err); ackErr != nil {
+				s.log.With("error", ackErr).Warnf("Failed to acknowledge path: %s", path)
+			}
+			continue
 		}
+
+		details := service.NewScannerSourceDetails()
+		details.SetName(path)
+		scanner, err := s.scannerCtor.Create(file, s.newCodecAckFn(path), details)
+		if err != nil {
+			_ = file.Close()
+			return fmt.Errorf("creating scanner: %w", err)
+		}
+
+		s.stateLock.Lock()
```

Comment:
Re-acquiring the lock here means there's a period of time where another goroutine could potentially have used the client to open a file, create a scanner and assign it, and now this goroutine is going to overwrite `s.scanner`, which means it's lost. I think in terms of performance you should be fine holding the lock for the entire method call.

Reply:
The problem holding the lock for the entire call is that in watch mode it will hold the lock open while `s.pathProvider.Next()` blocks and waits for a new file, which will cause AckFns to block because they need to grab the lock for a bit too to access `s.pathProvider`

Reply:
Ah I see 👍